File: ios/FinanceTracker/Networking/APIClient.swift
import Foundation

enum APIError: LocalizedError {
    case invalidURL
    case http(status: Int, detail: String?)
    case sessionExpired

    var errorDescription: String? {
        switch self {
        case .invalidURL: return "Invalid URL."
        case .http(let status, let detail): return detail ?? "Request failed (HTTP \(status))."
        case .sessionExpired: return "Your session has expired. Please log in again."
        }
    }
}

extension Notification.Name {
    static let sessionExpired = Notification.Name("sessionExpired")
}

/// Async URLSession client for the FastAPI backend.
/// Mirrors mobile/src/lib/api.ts: Bearer access token on every request,
/// transparent refresh-and-retry on 401 via /auth/refresh.
actor APIClient {
    static let shared = APIClient()

    /// The backend base URL. Debug builds allow a runtime override ("api_base_url" in
    /// UserDefaults, set from the More tab / login screen) so a physical device can reach
    /// a Mac on the LAN. Release builds ignore the override and use the baked-in endpoint.
    var baseURL: URL {
        #if DEBUG
        if let saved = UserDefaults.standard.string(forKey: "api_base_url"),
           !saved.isEmpty, let url = URL(string: saved) {
            return url
        }
        #endif
        return AppConfig.defaultBaseURL
    }

    // MARK: - JSON coding

    static let decoder: JSONDecoder = {
        let d = JSONDecoder()
        d.keyDecodingStrategy = .convertFromSnakeCase
        d.dateDecodingStrategy = .custom { decoder in
            let raw = try decoder.singleValueContainer().decode(String.self)
            if let date = DateParser.parse(raw) { return date }
            throw DecodingError.dataCorrupted(.init(
                codingPath: decoder.codingPath,
                debugDescription: "Unrecognized date format: \(raw)"
            ))
        }
        return d
    }()

    static let encoder: JSONEncoder = {
        let e = JSONEncoder()
        e.keyEncodingStrategy = .convertToSnakeCase
        e.dateEncodingStrategy = .formatted(DateParser.isoDateTime)
        return e
    }()

    // MARK: - Public API

    func get<T: Decodable>(_ path: String) async throws -> T {
        try await send(path: path, method: "GET", body: nil)
    }

    /// Raw (undecoded) GET — used for binary downloads like the CSV export ZIP.
    func getData(_ path: String) async throws -> Data {
        try await send(path: path, method: "GET", body: nil, decode: false)
    }

    func post<T: Decodable, B: Encodable>(_ path: String, body: B) async throws -> T {
        try await send(path: path, method: "POST", body: try Self.encoder.encode(body))
    }

    func put<T: Decodable, B: Encodable>(_ path: String, body: B) async throws -> T {
        try await send(path: path, method: "PUT", body: try Self.encoder.encode(body))
    }

    func patch<T: Decodable, B: Encodable>(_ path: String, body: B) async throws -> T {
        try await send(path: path, method: "PATCH", body: try Self.encoder.encode(body))
    }

    func delete(_ path: String) async throws {
        _ = try await send(path: path, method: "DELETE", body: nil, decode: false)
    }

    /// POST /auth/token with form-encoded credentials; stores both tokens.
    func login(email: String, password: String) async throws {
        var request = URLRequest(url: baseURL.appending(path: "/auth/token"))
        request.httpMethod = "POST"
        request.setValue("application/x-www-form-urlencoded", forHTTPHeaderField: "Content-Type")
        var components = URLComponents()
        components.queryItems = [
            URLQueryItem(name: "username", value: email),
            URLQueryItem(name: "password", value: password),
        ]
        request.httpBody = Data((components.percentEncodedQuery ?? "").utf8)

        let (data, response) = try await URLSession.shared.data(for: request)
        try Self.checkStatus(response, data: data)
        let token = try Self.decoder.decode(TokenResponse.self, from: data)
        Keychain.set(token.accessToken, for: Keychain.accessTokenKey)
        if let refresh = token.refreshToken {
            Keychain.set(refresh, for: Keychain.refreshTokenKey)
        }
    }

    // MARK: - Core request path

    @discardableResult
    private func send(
        path: String,
        method: String,
        body: Data?,
        decode: Bool = true,
        isRetry: Bool = false
    ) async throws -> Data {
        var request = URLRequest(url: Self.url(base: baseURL, path: path))
        request.httpMethod = method
        request.httpBody = body
        if body != nil {
            request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        }
        if let token = Keychain.get(Keychain.accessTokenKey) {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }

        let (data, response) = try await URLSession.shared.data(for: request)
        let status = (response as? HTTPURLResponse)?.statusCode ?? 0

        if status == 401, !isRetry, !path.hasPrefix("/auth/") {
            try await refreshAccessToken()
            return try await send(path: path, method: method, body: body, decode: decode, isRetry: true)
        }

        try Self.checkStatus(response, data: data)
        return data
    }

    /// Build a request URL from a path that may carry a query string.
    ///
    /// `URL.appending(path:)` percent-encodes its argument, so a path like
    /// "/x/projection?months=360" turns the "?" into "%3F" and the query becomes
    /// part of the path — the server sees an unknown route and 404s. Split the
    /// query off and attach it as real query items instead.
    static func url(base: URL, path: String) -> URL {
        guard let separator = path.firstIndex(of: "?") else {
            return base.appending(path: path)
        }
        let bare = String(path[path.startIndex..<separator])
        let query = String(path[path.index(after: separator)...])

        let items = query.split(separator: "&").compactMap { pair -> URLQueryItem? in
            let parts = pair.split(separator: "=", maxSplits: 1, omittingEmptySubsequences: false)
            guard let name = parts.first, !name.isEmpty else { return nil }
            return URLQueryItem(
                name: String(name),
                value: parts.count > 1 ? String(parts[1]) : nil
            )
        }

        let withPath = base.appending(path: bare)
        guard !items.isEmpty else { return withPath }
        return withPath.appending(queryItems: items)
    }

    private func send<T: Decodable>(path: String, method: String, body: Data?) async throws -> T {
        let data = try await send(path: path, method: method, body: body, decode: true)
        return try Self.decoder.decode(T.self, from: data)
    }

    private func refreshAccessToken() async throws {
        guard let refreshToken = Keychain.get(Keychain.refreshTokenKey) else {
            expireSession()
            throw APIError.sessionExpired
        }
        var request = URLRequest(url: baseURL.appending(path: "/auth/refresh"))
        request.httpMethod = "POST"
        request.setValue("Bearer \(refreshToken)", forHTTPHeaderField: "Authorization")
        do {
            let (data, response) = try await URLSession.shared.data(for: request)
            try Self.checkStatus(response, data: data)
            let token = try Self.decoder.decode(TokenResponse.self, from: data)
            Keychain.set(token.accessToken, for: Keychain.accessTokenKey)
            if let newRefresh = token.refreshToken {
                Keychain.set(newRefresh, for: Keychain.refreshTokenKey)
            }
        } catch {
            expireSession()
            throw APIError.sessionExpired
        }
    }

    private func expireSession() {
        Keychain.clearTokens()
        NotificationCenter.default.post(name: .sessionExpired, object: nil)
    }

    private static func checkStatus(_ response: URLResponse, data: Data) throws {
        guard let http = response as? HTTPURLResponse else { return }
        guard !(200...299).contains(http.statusCode) else { return }
        // FastAPI errors look like {"detail": "..."} (detail may also be a validation array).
        var detail: String?
        if let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] {
            detail = json["detail"] as? String
        }
        throw APIError.http(status: http.statusCode, detail: detail)
    }
}

/// Backend dates arrive as naive ISO datetimes ("2026-07-19T00:00:00[.ffffff]"),
/// timezone-aware ones ("2026-07-19T01:53:01.787071Z"), or plain dates ("2026-07-19").
///
/// Parsed by hand rather than by `DateFormatter`, because this is the app's hottest
/// decoding path: the Dashboard alone pulls balances, transactions and a daily portfolio
/// timeseries covering the household's whole history, which is tens of thousands of date
/// fields in one load. The old implementation tried five `DateFormatter`s in sequence and
/// fell back to a freshly-allocated `ISO8601DateFormatter`; a date-only string — the most
/// common shape the backend sends — cost four *failed* parses before the one that worked,
/// at roughly 100µs each. That is where the app's half-minute cold start went.
///
/// The scanner below reads fixed offsets out of the string's UTF-8 bytes and converts with
/// days-from-civil arithmetic, so there is no formatter, no `Calendar` and no allocation on
/// the happy path. `DateParser.legacyParse` keeps the old chain as a fallback for anything
/// the scanner doesn't recognise, so no format that used to decode stops decoding.
enum DateParser {
    /// Kept for *encoding* (`APIClient.encoder`) and as the parsing fallback.
    static let isoDateTime = formatter("yyyy-MM-dd'T'HH:mm:ss")
    static let isoDateTimeFractional = formatter("yyyy-MM-dd'T'HH:mm:ss.SSSSSS")
    static let isoDateTimeZoned = formatter("yyyy-MM-dd'T'HH:mm:ssXXXXX")
    static let isoDateTimeFractionalZoned = formatter("yyyy-MM-dd'T'HH:mm:ss.SSSSSSXXXXX")
    static let isoDateOnly = formatter("yyyy-MM-dd")

    static func parse(_ raw: String) -> Date? {
        fastParse(raw) ?? legacyParse(raw)
    }

    // MARK: - Fast path

    /// Reads `yyyy-MM-dd` optionally followed by `THH:mm:ss`, optional fractional seconds,
    /// and an optional `Z` / `±HH:MM` / `±HHMM` offset. Returns nil (rather than a wrong
    /// date) for anything else, including out-of-range components, so the fallback sees it.
    private static func fastParse(_ raw: String) -> Date? {
        // `withUTF8` gives a contiguous buffer without copying when the string already has
        // one, which is the case for every string `JSONDecoder` hands us.
        var raw = raw
        return raw.withUTF8(scan(_:))
    }

    private static func scan(_ bytes: UnsafeBufferPointer<UInt8>) -> Date? {
        let count = bytes.count
        guard count >= 10 else { return nil }

        // Bounds-checked here rather than at each call site: a truncated offset
        // ("…+8" instead of "…+08:00") otherwise reads past the buffer, which
        // `UnsafeBufferPointer` only traps on in debug builds.
        func digits(_ start: Int, _ length: Int) -> Int? {
            guard start >= 0, start + length <= count else { return nil }
            var value = 0
            for index in start..<(start + length) {
                let byte = bytes[index]
                guard byte >= 0x30, byte <= 0x39 else { return nil }
                value = value * 10 + Int(byte - 0x30)
            }
            return value
        }

        guard let year = digits(0, 4), bytes[4] == 0x2D,        // "-"
              let month = digits(5, 2), bytes[7] == 0x2D,
              let day = digits(8, 2),
              month >= 1, month <= 12, day >= 1, day <= 31
        else { return nil }

        var seconds = Double(daysFromCivil(year: year, month: month, day: day)) * 86_400

        guard count > 10 else {
            return Date(timeIntervalSince1970: seconds)
        }
        // Only "T" separates a date from a time here; a space-separated datetime never
        // parsed before either.
        guard bytes[10] == 0x54, count >= 19 else { return nil }
        guard let hour = digits(11, 2), bytes[13] == 0x3A,      // ":"
              let minute = digits(14, 2), bytes[16] == 0x3A,
              let second = digits(17, 2),
              hour <= 23, minute <= 59, second <= 60           // 60 = leap second, as ISO allows
        else { return nil }
        seconds += Double(hour * 3600 + minute * 60 + second)

        var index = 19
        // Fractional seconds: kept at whatever precision was sent, not truncated to the
        // six digits the old `.SSSSSS` formatter insisted on.
        if index < count, bytes[index] == 0x2E {               // "."
            index += 1
            let fractionStart = index
            var fraction = 0.0
            var scale = 1.0
            while index < count, bytes[index] >= 0x30, bytes[index] <= 0x39 {
                scale /= 10
                fraction += Double(bytes[index] - 0x30) * scale
                index += 1
            }
            guard index > fractionStart else { return nil }
            seconds += fraction
        }

        guard index < count else {
            return Date(timeIntervalSince1970: seconds)       // naive == UTC, as before
        }

        switch bytes[index] {
        case 0x5A, 0x7A:                                       // "Z" / "z"
            guard index == count - 1 else { return nil }
        case 0x2B, 0x2D:                                       // "+" / "-"
            let sign: Double = bytes[index] == 0x2B ? -1 : 1   // offset is removed to reach UTC
            let offsetStart = index + 1
            guard let offsetHour = digits(offsetStart, 2) else { return nil }
            let minuteStart = (offsetStart + 2 < count && bytes[offsetStart + 2] == 0x3A)
                ? offsetStart + 3
                : offsetStart + 2
            guard minuteStart + 2 == count, let offsetMinute = digits(minuteStart, 2) else { return nil }
            seconds += sign * Double(offsetHour * 3600 + offsetMinute * 60)
        default:
            return nil
        }
        return Date(timeIntervalSince1970: seconds)
    }

    /// Days between 1970-01-01 and the given proleptic-Gregorian date (Howard Hinnant's
    /// `days_from_civil`). Pure integer arithmetic — no `Calendar`, no timezone lookup.
    private static func daysFromCivil(year: Int, month: Int, day: Int) -> Int {
        let y = year - (month <= 2 ? 1 : 0)
        let era = (y >= 0 ? y : y - 399) / 400
        let yearOfEra = y - era * 400                                   // [0, 399]
        let dayOfYear = (153 * (month + (month > 2 ? -3 : 9)) + 2) / 5 + day - 1
        let dayOfEra = yearOfEra * 365 + yearOfEra / 4 - yearOfEra / 100 + dayOfYear
        return era * 146_097 + dayOfEra - 719_468
    }

    // MARK: - Fallback

    /// The original formatter chain, now only reached for shapes the scanner rejects.
    static func legacyParse(_ raw: String) -> Date? {
        isoDateTime.date(from: raw)
            ?? isoDateTimeFractional.date(from: raw)
            ?? isoDateTimeZoned.date(from: raw)
            ?? isoDateTimeFractionalZoned.date(from: raw)
            ?? isoDateOnly.date(from: raw)
            ?? sharedISO8601.date(from: raw)
    }

    /// One shared instance; the old code allocated an `ISO8601DateFormatter` per call.
    private static let sharedISO8601 = ISO8601DateFormatter()

    private static func formatter(_ format: String) -> DateFormatter {
        let f = DateFormatter()
        f.locale = Locale(identifier: "en_US_POSIX")
        f.timeZone = TimeZone(identifier: "UTC")
        f.dateFormat = format
        return f
    }
}

File: android/app/src/main/java/com/ivanlee/financetracker/data/net/ApiClient.kt
package com.ivanlee.financetracker.data.net

import android.content.Context
import android.content.SharedPreferences
import com.ivanlee.financetracker.BuildConfig
import com.ivanlee.financetracker.data.model.TokenResponse
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.withContext
// The reified extensions — without these imports the calls below bind to the
// (serializer, value) overloads and stop compiling.
import kotlinx.serialization.decodeFromString
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonNamingStrategy
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import okhttp3.HttpUrl
import okhttp3.HttpUrl.Companion.toHttpUrlOrNull
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import java.io.IOException
import java.util.concurrent.TimeUnit

/** Thrown for any non-2xx response; [detail] carries FastAPI's `{"detail": "..."}` message. */
class ApiException(val status: Int, val detail: String?) :
    IOException(detail ?: "Request failed (HTTP $status).")

/** The refresh token was missing or rejected — the user has to log in again. */
class SessionExpiredException : IOException("Your session has expired. Please log in again.")

/**
 * Coroutine HTTP client for the FastAPI backend.
 *
 * Mirrors ios/FinanceTracker/Networking/APIClient.swift: a Bearer access token on every
 * request and a transparent refresh-and-retry on 401 via `/auth/refresh`.
 */
object Api {

    /**
     * Shared JSON config. Three settings carry real weight:
     *  - [JsonNamingStrategy.SnakeCase] maps camelCase Kotlin names onto the Pydantic
     *    snake_case wire format (iOS's `.convertFromSnakeCase`).
     *  - `explicitNulls = false` omits null properties from request bodies, which is what
     *    makes every partial-update model ("send only what changed") work.
     *  - `ignoreUnknownKeys` keeps an older build decoding responses from a newer backend.
     */
    val json: Json = Json {
        namingStrategy = JsonNamingStrategy.SnakeCase
        ignoreUnknownKeys = true
        explicitNulls = false
        encodeDefaults = true
        isLenient = true
    }

    private val http: OkHttpClient = OkHttpClient.Builder()
        .connectTimeout(20, TimeUnit.SECONDS)
        .readTimeout(60, TimeUnit.SECONDS)
        .build()

    private val jsonMedia = "application/json".toMediaType()
    private val formMedia = "application/x-www-form-urlencoded".toMediaType()

    /** Emits once whenever the session dies, so the app can drop to the login screen. */
    val sessionExpired = MutableSharedFlow<Unit>(extraBufferCapacity = 1)

    private var appPrefs: SharedPreferences? = null

    fun init(context: Context) {
        TokenStore.init(context)
        appPrefs = context.applicationContext
            .getSharedPreferences("waypoint_prefs", Context.MODE_PRIVATE)
    }

    /**
     * The backend base URL.
     *
     * Debug builds allow a runtime override (`api_base_url`, set from Settings ▸ API server)
     * so a physical device can reach a Mac on the LAN; release builds ignore the override and
     * always use the baked-in endpoint. The whole override path is compiled out of release.
     */
    val baseUrl: String
        get() {
            if (BuildConfig.DEBUG) {
                val saved = appPrefs?.getString(OVERRIDE_KEY, null)
                if (!saved.isNullOrBlank()) return saved.trimEnd('/')
            }
            return BuildConfig.API_BASE_URL.trimEnd('/')
        }

    const val OVERRIDE_KEY = "api_base_url"

    /** Debug-only: point the app at a different backend (no-op in release). */
    fun setBaseUrlOverride(value: String?) {
        if (!BuildConfig.DEBUG) return
        appPrefs?.edit()?.apply {
            if (value.isNullOrBlank()) remove(OVERRIDE_KEY) else putString(OVERRIDE_KEY, value.trim())
        }?.apply()
    }

    fun currentOverride(): String? =
        if (BuildConfig.DEBUG) appPrefs?.getString(OVERRIDE_KEY, null) else null

    // ---- Public verbs -------------------------------------------------------------------

    suspend inline fun <reified T> get(path: String): T =
        json.decodeFromString(sendForString(path, "GET", null))

    /** Raw (undecoded) GET — used for binary downloads like the CSV export ZIP. */
    suspend fun getBytes(path: String): ByteArray = sendForBytes(path, "GET", null)

    suspend inline fun <reified B, reified T> post(path: String, body: B): T =
        json.decodeFromString(sendForString(path, "POST", json.encodeToString(body)))

    suspend inline fun <reified B, reified T> put(path: String, body: B): T =
        json.decodeFromString(sendForString(path, "PUT", json.encodeToString(body)))

    suspend inline fun <reified B, reified T> patch(path: String, body: B): T =
        json.decodeFromString(sendForString(path, "PATCH", json.encodeToString(body)))

    /** POST that returns a body we don't care about. */
    suspend inline fun <reified B> postIgnoringResult(path: String, body: B) {
        sendForBytes(path, "POST", json.encodeToString(body))
    }

    /** POST with no request body at all (action endpoints like `.../recurring/{id}/run`). */
    suspend inline fun <reified T> postEmpty(path: String): T =
        json.decodeFromString(sendForString(path, "POST", "{}"))

    suspend fun delete(path: String) {
        sendForBytes(path, "DELETE", null)
    }

    /** POST /auth/token with form-encoded credentials; stores both tokens. */
    suspend fun login(email: String, password: String) = withContext(Dispatchers.IO) {
        val form = "username=${email.formEncode()}&password=${password.formEncode()}"
        val request = Request.Builder()
            .url(url("/auth/token"))
            .post(form.toRequestBody(formMedia))
            .build()
        http.newCall(request).execute().use { response ->
            val body = response.body?.string().orEmpty()
            checkStatus(response.code, body)
            val token = json.decodeFromString<TokenResponse>(body)
            TokenStore.accessToken = token.accessToken
            token.refreshToken?.let { TokenStore.refreshToken = it }
        }
    }

    // ---- Core request path --------------------------------------------------------------

    suspend fun sendForString(path: String, method: String, body: String?): String =
        String(sendForBytes(path, method, body), Charsets.UTF_8)

    suspend fun sendForBytes(
        path: String,
        method: String,
        body: String?,
        isRetry: Boolean = false,
    ): ByteArray = withContext(Dispatchers.IO) {
        val builder = Request.Builder().url(url(path))
        val requestBody = body?.toRequestBody(jsonMedia)
        when (method) {
            "GET" -> builder.get()
            "DELETE" -> if (requestBody != null) builder.delete(requestBody) else builder.delete()
            else -> builder.method(method, requestBody ?: ByteArray(0).toRequestBody(jsonMedia))
        }
        TokenStore.accessToken?.let { builder.header("Authorization", "Bearer $it") }

        val response = http.newCall(builder.build()).execute()
        val bytes = response.use { it.body?.bytes() ?: ByteArray(0) }

        if (response.code == 401 && !isRetry && !path.startsWith("/auth/")) {
            refreshAccessToken()
            return@withContext sendForBytes(path, method, body, isRetry = true)
        }

        checkStatus(response.code, String(bytes, Charsets.UTF_8))
        bytes
    }

    private val refreshLock = Mutex()

    /**
     * Serialized behind [refreshLock] so a screen firing several requests at once doesn't
     * race three refreshes and have two of them invalidate the token the third just stored.
     */
    private suspend fun refreshAccessToken() = refreshLock.withLock {
        val refresh = TokenStore.refreshToken ?: run {
            expireSession()
            throw SessionExpiredException()
        }
        val request = Request.Builder()
            .url(url("/auth/refresh"))
            .header("Authorization", "Bearer $refresh")
            .post(ByteArray(0).toRequestBody(jsonMedia))
            .build()
        // A dropped connection, timeout, or server error here says nothing about whether the
        // refresh token itself is still valid — only a 401 from the server does. Wiping the
        // session on every transport failure turned a flaky network into a surprise full
        // sign-out; only an actual rejection should do that.
        http.newCall(request).execute().use { response ->
            if (response.code == 401) {
                expireSession()
                throw SessionExpiredException()
            }
            val body = response.body?.string().orEmpty()
            checkStatus(response.code, body)
            val token = json.decodeFromString<TokenResponse>(body)
            TokenStore.accessToken = token.accessToken
            token.refreshToken?.let { TokenStore.refreshToken = it }
        }
    }

    private fun expireSession() {
        TokenStore.clear()
        sessionExpired.tryEmit(Unit)
    }

    private fun checkStatus(code: Int, body: String) {
        if (code in 200..299) return
        // FastAPI errors look like {"detail": "..."} (detail may also be a validation array).
        val detail = runCatching {
            json.parseToJsonElement(body).jsonObject["detail"]?.jsonPrimitive?.content
        }.getOrNull()
        throw ApiException(code, detail)
    }

    /**
     * Build a request URL from a path that may carry a query string.
     *
     * OkHttp's `HttpUrl.Builder.addPathSegments` percent-encodes what it's given, so a path
     * like "/x/projection?months=360" would turn the "?" into "%3F" — the query becomes part
     * of the path and the server 404s. Split the query off and attach it as real query
     * parameters instead. (Same trap as iOS's `APIClient.url(base:path:)`; `ApiUrlTest` pins
     * the behaviour.)
     */
    fun url(path: String): HttpUrl {
        val base = baseUrl.toHttpUrlOrNull()
            ?: throw ApiException(0, "Invalid API base URL: $baseUrl")
        val separator = path.indexOf('?')
        val bare = if (separator < 0) path else path.substring(0, separator)
        val query = if (separator < 0) null else path.substring(separator + 1)

        val builder = base.newBuilder()
        bare.split('/').filter { it.isNotEmpty() }.forEach { builder.addPathSegment(it) }
        query?.split('&')?.forEach { pair ->
            if (pair.isEmpty()) return@forEach
            val eq = pair.indexOf('=')
            if (eq < 0) builder.addQueryParameter(pair, null)
            else builder.addQueryParameter(pair.substring(0, eq), pair.substring(eq + 1))
        }
        return builder.build()
    }

    private fun String.formEncode(): String =
        java.net.URLEncoder.encode(this, "UTF-8")
}
